### DISCUSSION TITLE
IKEv2 connection by SQL

### DISCUSSION THREAD

**Discussion by timofeyp** [strongswan/strongswan#1282] [Q&A]
Hello!
Im trying to confifure IKEv2 connection through sql, but can found specific tables for some properties. 
Here is my swanctl.cfg file, I mark some fields that cant locate in DB:

```
connections {
	ikev2-vpn {
		include /etc/swanctl/conf.d/ike_sa_default.conf    
		fragmentation=yes
		send_certreq=no
		rekey_time=0s
		version=2
		remote_addrs=%any
		local_addrs=%any
		send_cert=always
		proposals=chacha20poly1305-sha512-curve25519-prfsha512,aes256gcm16-sha384-prfsha384-ecp384,aes256-sha1-modp1024,aes128-sha1-modp1024,3des-sha1-modp1024
		pools=pool-subnet-10-10-10-0/24
		encap=yes
		dpd_delay=300s
		children {
			ikev2-vpn {
				include /etc/swanctl/conf.d/child_sa_default.conf
				start_action=none
				esp_proposals=chacha20poly1305-sha512,aes256gcm16-ecp384,aes256-sha256,aes256-sha1,3des-sha1
				local_ts=0.0.0.0/0 **<-----  WHERE THIS ?**
				dpd_action=clear
			}

		}
		local-0 { **<-----  WHERE THIS ?**
			certs = server-cert.pem
			id = 45.11.183.161
			eap_id=%identity
		}
		remote-0 { **<-----  WHERE THIS ?**
			auth = eap-mschapv2
			id = %any
		}
	}
}
pools {
	pool-subnet-10-10-10-0/24 {
		addrs=10.10.10.0/24
		dns=8.8.8.8,8.8.4.4
	}
}
secrets { 
	rsa-2 { **<-----  WHERE THIS ?**
		file="server-key.pem"
	}
	eap-user1 { **<-----  WHERE THIS ?**
		secret = "password"
		id-0=user1
	}
	eap-user2 { **<-----  WHERE THIS ?**
		secret = "password"
		id-0=user2
	}
	eap-user3 { **<-----  WHERE THIS ?**
		secret = "password"
		id-0=user3
	}
}
```

1. How should `connections.ikev2-vpn.children.ikev2-vpn.local_ts` be located ?
2. How should `connections.ikev2-vpn.children.ilocal-0` be located ?
3. How should `connections.ikev2-vpn.children.remote-0` be located ?
4. About secrets, `shared-secrets` table contain only `type` and `data`. How to correctly configure my secrets in DB? I guess secret field ===  `data` of `shared-secrets`. But which type ? Which field for secret name (eap-user1, eap-user2, eap-user3) ?  And how to configure rsa-2 secret and link it with key ? 

For another fields seems to be clear.

**Comment by tobiasbrunner** [comment]
See [here](https://docs.strongswan.org/docs/5.9/config/sqliteDbSchema.html) for more information on the tables (e.g. the type of secrets), and [here](https://docs.strongswan.org/docs/5.9/plugins/sql.html) for more information on the _sql_ plugin. There are also some utilities in the `scripts` directory to e.g. convert identities.

The traffic selectors are configured in a separate table (`traffic_selectors`) and then connected to the child configs via `child_config_traffic_selector` (see the data.sql example linked on the second page above). The authentication settings are specified in the `peer_configs` table, however, identities are stored in a separate table and referenced from there. But be aware that you currently can't configure `eap_id` via SQL.

Note that you don't have to configure the connection or the server's private key via SQL if you just want to manage user credentials in a database (i.e. you could just use the `identity` and `shared_secrets` tables with a config from swanctl.conf).
